### PR TITLE
Feature/state machine

### PR DESCRIPTION
**Pull request by kcyakici**

### CODE REVIEW

**Comment by kcyakici** [comment]
[Finite State Machine Implementation for the Player](https://trello.com/c/lzDYC5Of/73-finite-state-machine-implementation-for-the-player)

File: StateMachine/StateMachine.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+using Godot;
+using System;
+
+class StateMachine : Node2D
+{
+    private State currentState;
+
+    public override void _Input(InputEvent @event)
```
Comment (berkunal) [inline StateMachine/StateMachine.cs:8]: We may change this one to _UnhandledInput

File: StateMachine/StateMachine.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public void transition(String state)
+    {
+        if (HasNode(state))
+        {
+            State newState = GetNode<State>(state);
+            this.currentState = newState;
+            this.currentState.entry();
+            Owner.GetNode<Label>("CurrentState").Text = state;
```
Comment (berkunal) [inline StateMachine/StateMachine.cs:34]: We can make this label invisible when we are done dealing with the state machine

File: StateMachine/State.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+abstract class State : Node2D
+{
+    protected StateMachine parentStateMachine;
+    protected Player player;
+
+    public override void _Ready()
+    {
+        this.parentStateMachine = GetParent<StateMachine>();
+        this.player = (Player)Owner;
```
Comment (berkunal) [inline StateMachine/State.cs:12]: leave one space between casted var and the type -> `(Player) Owner`

File: Characters/Player/Player.cs [hunk 4]
```diff
@@ -3,12 +3,14 @@
 
 public class Player : KinematicBody2D
 {
-    private Vector2 velocity = new Vector2();
-    private int speed = 700;
-    private int jumpSpeed = -400;
-    private int maxSpeed = 1000;
-    private int gravity = 1000;
-    private float friction = 0.08f;
+    public Vector2 Velocity = new Vector2();
```
Comment (berkunal) [inline Characters/Player/Player.cs]: Apply encapsulation to this field also.

File: StateMachine/States/Air.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        // Apply gravity to the velocity vector
+        player.Velocity.y += player.Gravity * _delta;
+
+        // Move the instance
+        player.Velocity = player.MoveAndSlide(player.Velocity, Vector2.Up);
+    }
+
+    public override void handleInput(InputEvent @event)
```
Comment (berkunal) [inline StateMachine/States/Air.cs]: If you don't have a body you can declare a function in one line: `public override void handleInput(InputEvent @event) { }`

File: StateMachine/States/Air.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        // Update the velocity vector if user wants to move
+        if (player.DirectionOfMoving != 0)
+        {
+            player.Velocity.x += player.DirectionOfMoving * player.Speed * _delta;
+            player.Velocity.x = Mathf.Clamp(player.Velocity.x, -(player.MaxSpeed), player.MaxSpeed);
+        }
+
+        // Apply gravity to the velocity vector
```
Comment (berkunal) [inline StateMachine/States/Air.cs]: I don't think these comments are needed anymore.

File: StateMachine/States/Run.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    override public void physicsProcess(float _delta)
+    {
+        if (!player.IsOnFloor())
+        {
+            parentStateMachine.transition("Air");
+            return;
+        }
+
+        // Update the velocity vector
```
Comment (berkunal) [inline StateMachine/States/Run.cs]: I don't think these comments are needed anymore.

File: StateMachine/States/Idle.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            player.DirectionOfMoving = 1;
+            parentStateMachine.transition("Run");
+        }
+
+        else if (@event.IsActionPressed("move_left"))
+        {
+            player.DirectionOfMoving = -1;
+            parentStateMachine.transition("Run");
+        }
+
```
Comment (berkunal) [inline StateMachine/States/Idle.cs]: Delete empty line

File: StateMachine/States/Idle.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        player.DirectionOfMoving = 0;
+
+        player.Velocity.x = Mathf.Lerp(player.Velocity.x, 0, player.Friction);
+        player.Velocity.y += player.Gravity * _delta;
+        player.Velocity = player.MoveAndSlide(player.Velocity, Vector2.Up);
+
+        if (!player.IsOnFloor())
+            parentStateMachine.transition("Air");
+
```
Comment (berkunal) [inline StateMachine/States/Idle.cs]: Delete empty line

**Review by berkunal — CHANGES_REQUESTED** [review]
Very nice overall, only a few tweaks are needed.

**Review by berkunal — DISMISSED** [review]

**Review by berkunal — APPROVED** [review]

**Pull request merged**